src/Sulu/Component/Content/Document/Syncronization/DocumentRegistrator.php

```diff
@@ -0,0 +1,214 @@
+            }
+
+            $this->registerSingleDocumentWithPDM($propertyValue);
+        }
+
+        // TODO: Workaround for the fact that "parent" is not in the metadata,
+        // see: https://github.com/sulu-io/sulu-document-manager/issues/67
+        if ($document instanceof ParentBehavior) {
+            if ($parent = $document->getParent()) {
+                $this->registerSingleDocumentWithPDM($parent, true);
+            }
+        }
+    }
+
+    private function registerSingleDocumentWithPDM($document, $create = false)
+    {
+        $ddmInspector = $this->defaultManager->getInspector();
+        $pdmRegistry = $this->publishManager->getRegistry();
+
+        // if the PDM registry already has the document, then
+        // there is nothing to do - the document manager will
+        // handle the rest.
+        if (true === $pdmRegistry->hasDocument($document)) {
+            return;
+        }
+
+        // see if we can resolve the corresponding node in the PDM.
+        // if we cannot then we either return and let the document
+        // manager create the new node, or, if $create is true, create
+        // the missing node (this happens when registering a document
+        // which is a relation to the incoming DDM document).
+        if (false === $uuid = $this->resolvePDMUUID($document)) {
+            if (false === $create) {
+                return;
+            }
+
+            $this->publishManager->getNodeManager()->createPath(
+                $ddmInspector->getPath($document), $ddmInspector->getUuid($document)
+            );
+
+            return;
+        }
+
+        // register the DDM document against the PDM PHPCR node.
+        $node = $this->publishManager->getNodeManager()->find($uuid);
+        $locale = $ddmInspector->getLocale($document);
+        $pdmRegistry->registerDocument(
+            $document,
+            $node,
+            $locale
+        );
+    }
+
+    /**
+     * If possible, resolve the UUID of the node in the PDM corresponding to
+     * the DDM node.
+     *
+     * If the UUID does not exist, we check to see if the path exsits.
+     * if neither the path or UUID exist, then the PDM should create a new
+     * document and we ensure that the PARENT path exists and if it doesn't
+     * we syncronize the ancestor nodes from the DDM.
+     *
+     * We return FALSE in the case that a new document should be created.
+     *
+     * In the case the UUID does not exist we assume that in a valid system
+     * that path will also NOT EXIST. If the path does exist, then it means that
+     * the corresponding PHPCR nodes were created independently of each
+     * other and bypassed the syncrhonization system and we throw an exception.
+     *
+     * @throws \RuntimeException If the UUID could not be resolved and it would be
+     *                           invalid to implicitly allow the node to be created.
+     */
+    private function resolvePDMUUID($object)
```

Comment:
I will think about a name.

On Tue, Apr 19, 2016 at 08:33:36AM -0700, Daniel Rotter wrote:
>    In
>    [1]src/Sulu/Component/Content/Document/Syncronization/DocumentRegistrator.php:
> 
>  > +     * If the UUID does not exist, we check to see if the path exsits.
>  > +     * if neither the path or UUID exist, then the PDM should create a new
>  > +     * document and we ensure that the PARENT path exists and if it doesn't
>  > +     * we syncronize the ancestor nodes from the DDM.
>  > +     *
>  > +     * We return FALSE in the case that a new document should be created.
>  > +     *
>  > +     * In the case the UUID does not exist we assume that in a valid system
>  > +     * that path will also NOT EXIST. If the path does exist, then it means that
>  > +     * the corresponding PHPCR nodes were created independently of each
>  > +     * other and bypassed the syncrhonization system and we throw an exception.
>  > +     *
>  > +     * @throws \RuntimeException If the UUID could not be resolved and it would be
>  > +     *                           invalid to implicitly allow the node to be created.
>  > +     */
>  > +    private function resolvePDMUUID($object)
> 
>    I would even like createPathIfNotExists better, if you resist on using
>    createPath.
> 
>    —
>    You are receiving this because you authored the thread.
>    Reply to this email directly or [2]view it on GitHub
> 
>    Reverse link: [3]unknown
> 
> References
> 
>    Visible links
>    1. https://github.com/dantleech/sulu/pull/2#discussion_r60254145
>    2. https://github.com/dantleech/sulu/pull/2/files/d181f9007b19b02c9fc7f5bcb5786b99f3f92c89#r60254145
>    3. https://github.com/dantleech/sulu/pull/2/files/d181f9007b19b02c9fc7f5bcb5786b99f3f92c89#r60254145